tensorflow/core/kernels/data/tensor_dataset_op.cc

```diff
@@ -26,14 +26,35 @@ namespace {
 
 class TensorDatasetOp : public DatasetOpKernel {
  public:
-  explicit TensorDatasetOp(OpKernelConstruction* ctx) : DatasetOpKernel(ctx) {}
+  explicit TensorDatasetOp(OpKernelConstruction* ctx) : DatasetOpKernel(ctx) {
+    OP_REQUIRES_OK(ctx, ctx->GetAttr("Toutput_types", &output_types_));
+    OP_REQUIRES_OK(ctx, ctx->GetAttr("output_shapes", &output_shapes_));
+  }
 
   void MakeDataset(OpKernelContext* ctx, DatasetBase** output) override {
     OpInputList inputs;
     OP_REQUIRES_OK(ctx, ctx->input_list("components", &inputs));
-    // TODO(mrry): Validate that the shapes of the "components" tensors match
-    // the "shapes" attr.;
     std::vector<Tensor> components(inputs.begin(), inputs.end());
+    OP_REQUIRES(ctx, components.size() == output_shapes_.size(),
```

Comment:
You should use `VerifyTypesMatch` and `VerifyShapesCompatible` from https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/data/dataset_utils.h, instead of writing custom code.